pymeasure/instruments/velleman/velleman_k8090.py

```diff
@@ -0,0 +1,272 @@
+        # this works
+        return 1 + 0xFF - ((self.BYTE_STX + command + mask + param1 + param2) & 0xFF)
+
+    def switch_on_blocking(self, channels):
+        """Switch on a set of channels and wait for the confirmation.
+
+        See :attr:`switch_on`.
+
+        The set command is identical, and the processing part of
+        :attr:`status` is called.
+
+        :returns: See :attr:`status`
+        """
+        mask = _parse_channels(channels)
+        self.write(f"0x11,{mask}")
+        data = self.read()
+        items = [int(it) for it in data.split(",")]
+        return _get_process_status(items)
+
+    def switch_off_blocking(self, channels):
+        """Switch off a set of channels and wait for the confirmation.
+
+        See :attr:`switch_off`.
+
+        The set command is identical, and the processing part of
+        :attr:`status` is called.
+
+        :returns: See :attr:`status`
+        """
+        mask = _parse_channels(channels)
+        self.write(f"0x12,{mask}")
+        data = self.read()
+        items = [int(it) for it in data.split(",")]
+        return _get_process_status(items)
+
+    def write(self, command, **kwargs):
+        """The write command specifically for the protocol of the K8090.
+
+        This overrides the method from the ``Instrument`` class.
+
+        Each packet to the device is 7 bytes:
+
+        STX (0x04) - CMD - MASK - PARAM1 - PARAM2 - CHK - ETX (0x0F)
+
+        Where `CHK` is checksum of the package.
+
+        :param command: String like "CMD[, MASK, PARAM1, PARAM2]" - only CMD is mandatory
+        :type command: str
+        """
+        items_str = command.split(",")
+
+        items = [int(it, 16) for it in items_str]
+
+        cmd = items[0]
+        mask = items[1] if len(items) > 1 else 0
+        param1 = items[2] if len(items) > 2 else 0
+        param2 = items[3] if len(items) > 3 else 0
+
+        checksum = self._make_checksum(cmd, mask, param1, param2)
+
+        content = [
+            self.BYTE_STX,
+            cmd,
+            mask,
+            param1,
+            param2,
+            checksum,
+            self.BYTE_ETX,
+        ]
+
+        self.write_bytes(bytes(content))
+
+    def read(self, **kwargs):
+        """The read command specifically for the protocol of the K8090.
+
+        This overrides the method from the ``instrument`` class.
+
+        See :meth:`write`, replies from the machine use the same format.
+
+        A read will return a list of CMD, MASK, PARAM1 and PARAM2.
+        """
+        num_bytes = 7  # Each packet should always be exactly 7 bytes
+
+        if isinstance(self.adapter.connection, serial.Serial):
+            # Because we do not always read the reply to a setting, the buffer
+            # might not have been empty. Try to clear it completely but without waiting
+            # for more
+            num_bytes = max(7, self.adapter.connection.in_waiting)
```

Comment:
And what do we do with a VISAAdapter (default for pymeasure)?
You could use `read_bytes(-1)` to read all availabe bytes.

I see a problem here:
1. you switch on/off
2. the device responds with a message (buffer contains 7 bytes)
3. you send another command
4. you read (executed immediately afterwards). As there are already 7 bytes in the buffer you read those (wrong ones): ERROR
5. Now the device response arrives with the correct answer, but nobody reads it.

For that reason, it is better to put a `read_bytes(-1)` as `check_errors` method and set the `check_set_errors` parameter to True in the switch_on/off commands. Similar to this code (taken from hcp.TC038D, the read method does some error checking as well):
~~~python
    def check_errors(self):
        """To be called from the property setters to read the acknowledgment."""
        self.read()

    setpoint = Instrument.control(
        "R,0x106", "W,0x106,%i",
        """The setpoint of the oven in °C.""",
        check_set_errors=True,
        get_process=lambda v: v / 10,
        set_process=lambda v: int(round(v * 10)),
    )
~~~

Reply:
Oops, I didn't know the VISAadapter would work with a plain Serial interface. I'll try to test this later today. Kind of address should you submit to connect with e.g. `COM1`?

Reply:
Hm, you are right, that is a valid scenario. I'll implement it as you suggest.
What I dislike about `read_bytes(-1)` is it waits for the full timeout, which seems unnecessary. But maybe I can play with recognizing the STX/ETX bytes.

Reply:
> Oops, I didn't know the VISAadapter would work with a plain Serial interface. I'll try to test this later today. Kind of address should you submit to connect with e.g. COM1?

For full description, see https://pyvisa.readthedocs.io/en/latest/introduction/names.html
How you address the device, depends on the backend you use: from NI or keysight or pyvisa-py. The defined way (which should work always) is "ASRL1". The NI backend also understands "COM1". The pyvisa-py backend, however, does not understand "COM1" and requires "ASRL1".

> What I dislike about read_bytes(-1) is it waits for the full timeout, which seems unnecessary. But maybe I can play with recognizing the STX/ETX bytes.

The SerialAdapter waits for the timeout. The VISAAdapter, however, recognizes STX/ETX (if I understand it correctly) out of the box.

Reply:
Yeah. I'm making some progress with a new implementation.

I am done with a version that uses the ETX byte to prevent most unnecessary waits, but guess what: the K8090 will also send the status when the test button on-board are used... So we will always need to drain the buffer fully, we can't even rely on the terminators.
I am getting real tired of this device :')

Reply:
Ok...

So we're back at handling it in "read".
What do you think: You read all available bytes and then you use the last 7 bytes only?

Reply:
Pff, pushed a new update with a new idea:

* before we do a write we just clear the buffer, to make sure no old messages are present.
* Then we write 
* and in `set_check_errors` we read the reply (if any). Reading is assisted by the ETX byte to speed up the process in all normal use cases.

Right now we must read by the ETX byte because #863 .

Reply:
@bmoneke I am resolving this conversation, I think the current implementation is good.